src/PowerPing/Ping.cs

```diff
@@ -410,24 +417,39 @@ private PingResults SendICMP(PingAttributes attrs)
                         if (rnd.Next(20) == 1) { throw new SocketException(); }
                     }
 
-                    // Wait for response
-                    byte[] buffer = new byte[attrs.RecieveBufferSize]; // Ipv4Header.length + IcmpHeader.length + attrs.recievebuffersize
-                    bytesRead = sock.ReceiveFrom(buffer, ref ep);
-                    responseTimer.Stop();
-
-                    // Store reply packet
-                    ICMP response = new ICMP(buffer, bytesRead);
+                    ICMP response;
+                    TimeSpan replyTime;
+                    do {
+                        // Wait for response
+                        bytesRead = sock.ReceiveFrom(receiveBuffer, ref ep);
+                        replyTime = new TimeSpan(Helper.StopwatchToTimeSpanTicks(Stopwatch.GetTimestamp() - requestTimestamp));
+
+                        // Store reply packet
+                        response = new ICMP(receiveBuffer, bytesRead);
+
+                        // Ignore unexpected echo responses
+                        if (packet.type == 8 && response.type == 0) {
+                            ushort responseSessionId = BitConverter.ToUInt16(response.message, 0);
+                            ushort responseSequenceNum = BitConverter.ToUInt16(response.message, 2);
+                            if (responseSessionId != sessionId || responseSequenceNum != sequenceNum) {
+                                if (replyTime.TotalMilliseconds >= attrs.Timeout) {
+                                    throw new SocketException();
+                                }
+                                response = null;
+                            }
+                        }
+                    } while (response == null);
 
                     // Display reply packet
                     if (Display.ShowReplies) {
-                        PowerPing.Display.ReplyPacket(response, Display.UseInputtedAddress | Display.UseResolvedAddress ? attrs.Host : ep.ToString(), index, responseTimer.Elapsed, bytesRead);
+                        PowerPing.Display.ReplyPacket(response, Display.UseInputtedAddress | Display.UseResolvedAddress ? attrs.Host : ep.ToString(), index, replyTime, bytesRead);
                     }
```

Comment:
In some situations we might be displaying packets that aren't either type 0 or 8, so maybe either the check on line 431 should change like I mentioned above.